src/Proxy.NET/Services/CatalogService.cs

```diff
@@ -0,0 +1,109 @@
+using System.Data;
+using System.Net;
+using AzureOpenAIProxy.Management;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Caching.Memory;
+using Npgsql;
+using Proxy.NET.Models;
+
+namespace Proxy.NET.Services;
+
+public class CatalogService(AoaiProxyContext db, IConfiguration configuration, IMemoryCache memoryCache) : ICatalogService
+{
+    private readonly string EncryptionKey =
+        configuration["PostgresEncryptionKey"] ?? throw new ArgumentNullException("PostgresEncryptionKey");
+
+    /// <summary>
+    /// Retrieves the event catalog for a given event ID and deployment name.
+    /// Calls function aoai.get_models_by_deployment_name that decrypts the endpoint URL and Key.
+    /// </summary>
+    /// <param name="eventId">The ID of the event.</param>
+    /// <param name="deploymentName">The name of the deployment.</param>
+    /// <returns>A list of Deployment objects representing the event catalog.</returns>
+    private async Task<List<Deployment>> GetDecryptedEventCatalogAsync(string eventId, string deploymentName)
+    {
+        if (memoryCache.TryGetValue(eventId + deploymentName, out List<Deployment>? cachedContext))
+            return cachedContext!;
+
+        var result = await db.Set<Deployment>()
+            .FromSqlRaw(
+                "SELECT * FROM aoai.get_models_by_deployment_name(@eventId, @deploymentName, @encryptionKey)",
+                new NpgsqlParameter("@eventId", eventId),
+                new NpgsqlParameter("@deploymentName", deploymentName),
+                new NpgsqlParameter("@encryptionKey", EncryptionKey)
+            )
+            .ToListAsync();
+
+        memoryCache.Set(eventId + deploymentName, result, TimeSpan.FromMinutes(2));
+        return result;
+    }
+
+    /// <summary>
+    /// Retrieves a catalog item.
+    /// </summary>
+    /// <param name="requestContext">The authorization response containing the event ID and deployment name.</param>
+    /// <returns>The deployment associated with the specified catalog name.</returns>
+    public async Task<Deployment> GetCatalogItemAsync(RequestContext requestContext)
+    {
+        var deployments = await GetDecryptedEventCatalogAsync(requestContext.EventId, requestContext.DeploymentName);
+
+        if (deployments.Count == 0)
+        {
+            throw new HttpRequestException(
+                $"The requested model '{requestContext.DeploymentName}' is not available for this event. "
+                    + $"Available models are: {string.Join(", ", (await GetEventCatalogAsync(requestContext.EventId)).Select(d => d.DeploymentName))}",
+                null,
+                HttpStatusCode.BadRequest
+            );
```

Comment:
The service shouldn't throw a `HttpRequestException`, as it should be unaware of the underlying protocol being used. Either this method should return `null`, or be in a `TryGet` design and return `(bool, Deployment?)`.

Also, we shouldn't put the available models in the exception message, partially because it means an extra DB hit, but it also means we're (potentially) leaking info.

Reply:
ahhh you've removed the discoverablity of deployments for attendees esp those doing hacks. This is needed as I see in the logs people trying models and they will never know what models are available. This needs to go back in :)

Reply:
This shouldn't be the responsibility of the API to return that information, it should be on the event to have that clearly documented, putting it in the API response provides a leaking of information.

Reply:
We know this has to be more discoverable, the 2500 dev event atm, relies on this discoverability, it just makes it out of the box and works well. I can tell from what I see in the logs for the current Python proxy...